### ISSUE TITLE
fix(sampling): apply top-p / min-p / XTC on the untempered distribution and scale by temperature last

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1379]
## Summary

mlxcel divides the logits by the temperature before top-k, top-p and min-p run, and applies XTC before any of them on the raw distribution. The llama-server sampler chain that mlxcel's flags and request schema are designed to be compatible with does the opposite: every truncation filter (top-k, typical-p, top-p, min-p, XTC) is evaluated on the untempered distribution and the temperature is applied last, only to the draw. With the server defaults (`--temp 0.8 --top-k 40 --top-p 0.9 --min-p 0.1`, `src/main.rs:1254-1267`) every default request is affected: at `T = 0.8` the tempered distribution is sharper, so top-p 0.9 and min-p 0.1 keep fewer tokens than the same request keeps on llama-server, and the nucleus a client tuned elsewhere does not transfer. This issue moves the temperature scaling to the end of the chain and XTC after min-p, in the Rust pre-step, the C++ stock chain, the rejection kernel, and the #902 distribution helper, so the sampled support matches the reference chain for every combination of the existing knobs.

## Current behavior

Rust pre-step applies XTC first, on the raw, untempered, unfiltered logits:

```rust
// src/lib/mlxcel-core/src/sampling.rs:416-426
// XTC (Exclude Top Choices): a logits pre-processing step, applied here
// (before the fused temperature/top-k/top-p/min-p/categorical sampler)
if config.xtc_probability > 0.0 { apply_xtc_step(&last_logits, config) } else { last_logits }
```

The C++ stock chain scales by temperature before any filter:

```cpp
// src/lib/mlxcel-core/cpp/mlx_cxx_bridge.cpp:5066-5068
// Temperature scaling
if (temperature > 0.0f && temperature != 1.0f) { x = x / sampler_scalar(temperature); }
// ... 5101-5105: top_k_filter(x), then top_p_filter(x, ...), 5109: compiled min-p
```

The rejection kernel (#901) receives a single probability vector that is already tempered, and evaluates the top-k count test, the top-p mass test and the min-p threshold against it:

```cpp
// src/lib/mlxcel-core/cpp/mlx_cxx_bridge.cpp:5270-5279
static mlx::core::array rejection_probabilities(const mlx::core::array& logits, float temperature) {
    auto x = mlx::core::astype(logits, mlx::core::float32);
    if (temperature > 0.0f && temperature != 1.0f) { x = mlx::core::divide(x, mlx::core::array(temperature)); }
    return mlx::core::softmax(x, -1);
}
```

`fused_sample_probs` (`mlx_cxx_bridge.cpp:5650-5680`) reports `softmax(filter_chain(x / T))`, so the speculative accept test inherits the same ordering. Effective order today:

```
XTC(raw) -> temperature -> top_k -> top_p (renormalised) -> min_p -> draw
```

Reference (llama-server) order for the same knobs:

```
top_k -> top_p -> min_p -> XTC -> temperature -> draw
```

Concrete divergence: row `p = [0.50, 0.30, 0.15, 0.05]`, `T = 0.5`, `top_p = 0.9`. Untempered nucleus keeps three tokens (`0.50 + 0.30 = 0.80 < 0.9`, third token enters). Tempered `p^2 / Z = [0.685, 0.247, 0.062, 0.007]`: `0.685 + 0.247 = 0.932 >= 0.9`, so only two tokens survive. Same request, different support.

## Expected behavior

Chain order, for every sampler entry point (`sample_token_optimized`, `batched_fused_sample`, `sample_token_with_distribution`, `effective_token_distribution`):

```
token bias -> repetition -> DRY -> frequency/presence
  -> [row filters from #1375, unchanged]
  -> top_k -> top_p -> min_p         (all evaluated on softmax(x), T not applied)
  -> XTC                             (threshold test on softmax of the filtered row, i.e. renormalised after top_p / min_p)
  -> x_filtered / T -> categorical   (temperature applied once, at the end)
```

- Greedy (`T == 0` or `top_k == 1`): argmax of the penalized logits, unchanged, byte-identical to today.
- `T == 1.0`: identical support and identical draws to today for every knob except XTC placement (XTC now sees the renormalised post-filter distribution).
- Relative order of top-k and top-p is NOT changed by this issue: the stock chain keeps "top-k, then top-p on the renormalised top-k set" and the rejection kernel keeps its documented intersection semantics (`mlx_cxx_bridge.cpp:5015-5037`). Both are evaluated on the untempered row after this change.
- `fused_sample_probs` returns `softmax(x_filtered / T)` where the filters were evaluated on `softmax(x)`, so the #902 proposal and target distributions follow the sampler.
- `min_p` stays invariant (its test `p_i >= min_p * p_max` is renormalisation-invariant, but it is temperature-variant, so it must now run on the untempered row like the others).

## Implementation plan

1. `src/lib/mlxcel-core/cpp/mlx_cxx_bridge.cpp`, `fused_sample_filter_logits` (5039-5125): delete the temperature block at 5066-5068; keep top-k / top-p / min-p exactly as they are (they now see the raw `x`); add the division by `sampler_scalar(temperature)` as the last statement before `return x`, guarded by `temperature > 0 && temperature != 1`. `fused_sample_argpartition_chain` and `fused_sample_probs` need no change beyond that (they call this function), but update the `fused_sample_probs` header comment that says the filters run on `x / T`.
2. Rejection kernel path, same file:
   - `rejection_probabilities` (5270-5279): split into `rejection_filter_probabilities(logits)` = `softmax(astype(logits, f32))` and `rejection_draw_probabilities(logits, T)` = `softmax(astype(logits, f32) / T)`; return the same array object for both when `T == 1.0` (no second softmax).
   - `src/lib/mlx-cpp/turbo/sampling_rejection.{h,cpp,metal source}`: `rejection_sample(probs_filter, probs_draw, params, max_rounds)`. The row-maximum sweep, the `total` mass, `tau_min = min_p * p_max`, `pmass_target = top_p * total`, and the `(count, mass)` reductions above `pivot_0` / `pivot_1` all read `probs_filter`. The proposal-mass scan, the per-thread exclusive CDF offsets, and `target = u * mass_prop` read `probs_draw`, restricted to the proposal set `{probs_filter > low}`. The accepted candidate's pivot value `pivot_0 = probs_filter[candidate]`. Correctness: `probs_draw_i = probs_filter_i^(1/T) / Z`, a strictly monotone map, so a threshold set in filter space is the same index set as a threshold set in draw space and the bracket / bisection argument in the header comment is unchanged; the accepted draw is distributed as `probs_draw` truncated to the filtered support, which is the target. Update the CUDA port identically (`bitlinear`-style mirrored source at `sampling_rejection.cpp:375+`).
   - `fused_sample_rejection_path`, `fused_sample_rejection`, `fused_sample_rejection_deferred`, `sampling_rejection_probe` (5555-5620): pass both arrays. Keep the kernel's `template_args` dtype-keyed (`make verify-kernel-dtype-keys`).
   - Determinism contract (header comment "Determinism"): the sampled id is now a pure function of `(key, probs_filter, probs_draw, params)`; update the comment.
3. XTC placement: the XTC threshold test needs the post-min-p renormalised distribution, which lives in C++. Move it into the stock chain: extend `fused_sample_filter_logits` (and the two `fused_sample*` entry points plus `fused_sample_probs`) with `xtc_threshold: f32`, `xtc_probability: f32`, `xtc_special_tokens: const MlxArray*` (nullable, int32 `[n]`), and `xtc_enabled: bool`. After min-p: `probs = softmax(x)`; `above = probs > threshold`; if `count(above) >= 2`, remove every `above` position except the least probable one and except the allowlist; gate with one `random::uniform()` draw against `xtc_probability` exactly as `apply_xtc_filter` / `apply_xtc_step` do today (`sampling.rs:1337-1430`). When XTC is enabled, `fused_sample_impl` skips the rejection kernel and takes the stock chain (XTC already forces the per-row path through `config_supports_fused_batch`, so no batch-path change). In Rust, delete the XTC block from `preprocess_logits_for_sampling` (421-426) and thread `config.xtc_*` into the four `ffi::fused_sample*` call sites (`sample_token_optimized_core`, `effective_token_distribution`, `sample_token_with_distribution`, `batched_fused_sample` passes disabled). Keep `apply_xtc_filter` as the Rust reference for the unit tests and make the C++ implementation match it element-for-element.
4. Rust reference filters `top_p_filter` / `min_p_filter` / `top_k_filter` (`sampling.rs:1219-1326`) are temperature-free already; no change. Update the `sample_token_optimized` and `preprocess_logits_for_sampling` doc comments that describe the old order.
5. `docs/`: document the chain order in the sampling section (one line, current state only).

## Validation

(a) Unit tests (append to the `sampling.rs` `mod tests`, `src/lib/mlxcel-core/src/sampling_rejection_tests.rs`, `src/lib/mlxcel-core/src/speculative/distribution_tests.rs`):

- `top_p_support_is_temperature_invariant`: row `log([0.50, 0.30, 0.15, 0.05])`, `top_p = 0.9`, `T in {0.5, 1.0, 2.0}`: `fused_sample_probs` has non-zero mass on exactly indices `{0, 1, 2}` for all three.
- `min_p_support_is_temperature_invariant`: same shape with `min_p = 0.2`: support `{0, 1}` for all `T`.
- `fused_sample_probs_equals_softmax_of_filtered_over_t`: for random rows and `T = 0.7`, `fused_sample_probs` equals `softmax(where(support, x, -inf) / T)` where `support` is computed by the Rust reference filters on the untempered row, within `1e-5`.
- `rejection_kernel_filter_on_untempered_draw_on_tempered`: 100k draws from the rejection path on a fixed row at `T = 0.5`, `top_p = 0.9`: empirical frequencies match `softmax(x_support / T)` within 3 sigma, and no draw lands outside the untempered nucleus. Run under `MLXCEL_SAMPLING_REJECTION=0` as well and require the same support (stock chain).
- `temperature_one_support_unchanged`: for `T = 1.0` and random `(top_k, top_p, min_p)` without XTC, `fused_sample_probs` is bit-identical before and after (regression guard on the snapshot committed with this change).
- `greedy_path_unchanged`: `T = 0` and `top_k = 1` outputs equal the argmax for random rows.
- `xtc_runs_on_renormalised_filtered_row`: row where top-p removes a token with `p > xtc_threshold`; after the change that token does not count towards XTC's "at least two above threshold" test.
- `stochastic_accept_uses_untempered_support`: the #902 accept test with `T = 0.5`, `top_p = 0.9` never rejects a target token inside the untempered nucleus on account of zero reported mass.

(b) Real checkpoint: `mlx-community/Qwen3-4B-4bit`.

```
./target/release/mlxcel generate -m models/Qwen3-4B-4bit -p "List five uses for a paperclip." -n 64 --temp 0.8 --top-k 40 --top-p 0.9 --min-p 0.1 --seed 7
./target/release/mlxcel generate -m models/Qwen3-4B-4bit -p "List five uses for a paperclip." -n 64 --temp 0 
```

Acceptance: greedy output is token-identical to the pre-change binary; the sampled run is fluent and finite; a 2k-token decode benchmark (`bench_decode`) at the server defaults shows no throughput regression beyond noise (the rejection kernel reads one extra `[B, V]` f32 row per launch when `T != 1`; report the measured delta in the PR). Speculative decoding (`--draft-model` with a Qwen3-0.6B drafter) at `T = 0.8` keeps its acceptance rate within 2 points of the pre-change build on the same prompt set.

## Acceptance criteria

- [x] Temperature is applied after top-k / top-p / min-p / XTC on the stock chain, the rejection kernel path, and `fused_sample_probs`
- [x] XTC evaluates on the renormalised post-min-p distribution and still forces the per-row path
- [x] Greedy output byte-identical; `T == 1.0` support byte-identical for non-XTC configs
- [x] `MLXCEL_SAMPLING_REJECTION=0` and the default path produce the same support for every configuration in the tests
- [x] Rejection kernel `template_args` stay dtype-keyed; `make verify-kernel-dtype-keys` passes
- [x] Unit tests listed above pass; real-checkpoint and benchmark checks above recorded in the PR
- [ ] cargo test --workspace --profile test-fast --features metal,accelerate passes
- [ ] cargo clippy --workspace --all-targets -- -D warnings and cargo fmt --all -- --check pass

## Out of scope

- Changing the relative order of top-k and top-p, or unifying the stock chain's renormalised-top-p with the rejection kernel's intersection semantics (documented in `mlx_cxx_bridge.cpp:5015-5037`).
- `min_tokens_to_keep` for min-p (not exposed today; unrelated).
- The new row filters (`top_n_sigma`, `typical_p`, `p_less`); they are defined on the untempered row (or take `T` explicitly) and sit before this chain either way.
